Havit.Blazor.Components.Web.Bootstrap/Forms/Internal/HxMultiSelectInternal.razor

```diff
@@ -36,25 +34,60 @@
     <ul class="dropdown-menu"> @* Must be always rendered otherwise does not work after disable->enabled scenario *@
         @if (enabled)
         {
-            for (int i = 0; i < ItemsToRender.Count; i++)
+            if (AllowFiltering)
+            {
+                <input @ref="filterInputReference"
+                       id="@($"{InputId}_filter")"
+                       type="text"
+                       class="@CssClassHelper.Combine("form-control")"
+                       autocomplete="false"
+                       value="@filterText"
+                       @oninput="HandleInputChangedAsync"
+                       @onclick:stopPropagation
+                       onfocusin="this.select()" />
+            }
+
+            var filteredItems = GetFilteredItems();
+            if (filteredItems.Count <= 0)
+            {
+                @EmptyTemplate
+            }
+            else
             {
-                string checkboxElementId = InputId + "_" + i.ToString();
-
-                var item = ItemsToRender[i];
-                TValue value = SelectorHelpers.GetValue<TItem, TValue>(ValueSelector, item);
-
-                bool itemSelected = Value?.Contains(value) ?? false;
-
-                <li>
-                    <button type="button" class="dropdown-item" role="option" @onclick="async () => await HandleItemSelectionChangedAsync(!itemSelected, item)">
-                        <div class="form-check">
-                            <input class="form-check-input" type="checkbox" id="@checkboxElementId" checked="@itemSelected" tabindex="-1" />
-                            <label class="form-check-label" for="@checkboxElementId" @onclick:preventDefault >
-                                @SelectorHelpers.GetText(TextSelector, item)
-                            </label>
-                        </div>
-                    </button>
-                </li>
+                if (AllowSelectAll)
+                {
+                    <li>
+                        <button type="button" class="dropdown-item" role="option" @onclick="HandleSelectAllClickedAsync">
+                            <div class="form-check">
+                                <input class="form-check-input" type="checkbox" id="@($"{InputId}_selectall")" checked="@selectAll" tabindex="-1" />
+                                <label class="form-check-label" for="@($"{InputId}_selectall")" @onclick:preventDefault>
+                                    @(SelectAllText ?? "-- select all --")
```

Comment:
We have to provide localization for the default text "select all" from the `IStringLocalizer<>`.

Reply:
I've reworked this to call a helper method which uses the newly injected IStringLocalizer with new resource file. I've only added English for now as that's all I'm comfortable with.